3.1.0 news (#11694)
dmlc/xgboost: doc/changes/v3.1.0.rst

###################
3.1.0 (2025 Sep 22)
###################

We are delighted to share the latest 3.1.0 update for XGBoost.

********************
Categorical Re-coder
********************

This release features a major update to categorical data support by introducing a
re-coder. This re-coder saves categories in the trained model and re-codes the data during
inference, to keep the categorical encoding consistent. Aside from primitive types like
integers, it also supports string-based categories. The implementation works with all
supported Python DataFrame implementations. (:pr:`11609`, :pr:`11665`, :pr:`11605`,
:pr:`11628`, :pr:`11598`, :pr:`11591`, :pr:`11568`, :pr:`11561`, :pr:`11650`, :pr:`11621`,
:pr:`11611`, :pr:`11313`, :pr:`11311`, :pr:`11310`, :pr:`11315`, :pr:`11303`, :pr:`11612`,
:pr:`11098`, :pr:`11347`) See :ref:`cat-recode` for more information. (:pr:`11297`)

> In addition, categorical support for polars is now available (:pr:`11565`).

Reviewer: ```suggestion
In addition, categorical support for Polars data frames is now available (:pr:`11565`).
```

Lastly, we removed the experimental tag for categorical feature support in this
release. (:pr:`11690`)

***************
External Memory
***************

We continue the work on external memory support on 3.1. In this release, XGBoost features
an adaptive cache for CUDA external memory. The improved cache can split the data between
CPU memory and GPU memory according to the underlying hardware and data
size. (:pr:`11556`, :pr:`11465`, :pr:`11664`, :pr:`11594`, :pr:`11469`, :pr:`11547`,
:pr:`11339`, :pr:`11477`, :pr:`11453`, :pr:`11446`, :pr:`11458`, :pr:`11426`, :pr:`11566`,
:pr:`11497`)

Also, there's an optional support (opt-in) for using ``nvcomp`` and the GB200
decompression engine to handle sparse data (requires nvcomp as a plugin) (:pr:`11451`,
:pr:`11464`, :pr:`11460`, :pr:`11512`, :pr:`11520`). We improved the memory usage of
quantile sketching with external memory (:pr:`11641`) and optimized the predictor for
training (:pr:`11548`). To help ensure the training performance, the latest XGBoost

> features detection for NUMA node (:pr:`11538`, :pr:`11576`) for checking cross-socket data

Reviewer: ```suggestion
features detection for NUMA (Non-Uniform Memory Access) node (:pr:`11538`, :pr:`11576`) for checking cross-socket data
```

- [sycl] Reduce predictor memory consumption and improve L2 locality (:pr:`11603`)

*****
Fixes
*****

- Fix static linking C++ libraries on macOS (:pr:`11522`)
- Rename param.hh/cc to hist_param.hh/cc to fix xcode build (:pr:`11378`)
- [sycl] Fix build with updated compiler (:pr:`11618`)
- [sycl] Various fixes for fp32-only devices. (:pr:`11527`, :pr:`11524`)
- Fix compilation on android older than API 26 (:pr:`11366`)
- Fix loading Gamma model from 1.3. (:pr:`11377`)

**************
Python Package
**************

- Support mixing Python metrics and built-in metrics for the skl interface. (:pr:`11536`)
- CUDA 13 Support for PyPI with the new ``xgboost-cu13`` package. (:pr:`11677`, :pr:`11662`)
- Remove wheels for manylinux2014. (:pr:`11673`)
- Initial support for building variant wheels (:pr:`11531`, :pr:`11645`, :pr:`11294`)
- Minimum PySpark version is now set to 3.4 (:pr:`11364`). In addition, the PySpark
  interface now checks the validation indicator column type and has a fix for None column
  input. (:pr:`11535`, :pr:`11523`)
- [dask] Small cleanup for the predict function. (:pr:`11423`)

*********
R Package
*********

> With most of the deprecated features have been removed in this release. We will try to

Reviewer: ```suggestion
Now that most of the deprecated features have been removed in this release, we will try to
```
